From 54b1518d4199ba22cd48f6ff7565561a0179d22e Mon Sep 17 00:00:00 2001
From: Lionel Debieve <lionel.debieve@foss.st.com>
Date: Tue, 1 Feb 2022 13:04:01 +0100
Subject: [PATCH 0205/1695] plat-stm32mp1: plat_tzc400: move driver to
 dt_driver

Change the tzc400 driver to register as a dt_generic_driver.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 327533af03d81fe3b23a1dc9f7775cd3ae7e32bf plat-stm32mp1: plat_tzc400: move driver to dt_driver]
Change-Id: I034045c2b138a9a67583873f6b2027f78d851e8b
Signed-off-by: Lionel Debieve <lionel.debieve@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/plat_tzc400.c | 174 +++++++++++++++++++++-
 1 file changed, 169 insertions(+), 5 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/plat_tzc400.c b/core/arch/arm/plat-stm32mp1/plat_tzc400.c
index 95ac0e023..7919291af 100644
--- a/core/arch/arm/plat-stm32mp1/plat_tzc400.c
+++ b/core/arch/arm/plat-stm32mp1/plat_tzc400.c
@@ -5,10 +5,15 @@
 
 #include <assert.h>
 #include <config.h>
+#include <drivers/clk.h>
+#include <drivers/clk_dt.h>
 #include <drivers/tzc400.h>
 #include <initcall.h>
+#include <io.h>
+#include <kernel/dt.h>
 #include <kernel/interrupt.h>
 #include <kernel/panic.h>
+#include <libfdt.h>
 #include <mm/core_memprot.h>
 #include <platform_config.h>
 #include <trace.h>
@@ -21,6 +26,35 @@
 #define TZC_FILTERS_MASK	GENMASK_32(0, 0)
 #endif
 
+struct stm32mp_tzc_region {
+	uint32_t cfg;
+	uint32_t addr;
+	uint32_t len;
+};
+
+struct stm32mp_tzc_platdata {
+	const char *name;
+	uintptr_t base;
+	struct clk *clk[2];
+	int irq;
+	uint32_t mem_base;
+	uint32_t mem_size;
+	struct firewall_compat *tzc_compat;
+	struct stm32mp_tzc_region *regions;
+};
+
+struct stm32mp_tzc_driver_data {
+	uint32_t nb_filters;
+	uint32_t nb_regions;
+};
+
+struct tzc_device {
+	struct stm32mp_tzc_platdata pdata;
+	struct stm32mp_tzc_driver_data *ddata;
+	struct tzc_region_config *reg;
+	bool *reg_locked;
+};
+
 static enum itr_return tzc_it_handler(struct itr_handler *handler __unused)
 {
 	EMSG("TZC permission failure");
@@ -104,13 +138,126 @@ static void stm32mp_tzc_check_boot_region(void)
 	}
 }
 
-static TEE_Result init_stm32mp1_tzc(void)
+static void tzc_set_driverdata(struct tzc_device *tzc_dev)
+{
+	struct stm32mp_tzc_driver_data *ddata = tzc_dev->ddata;
+	uintptr_t base = tzc_dev->pdata.base;
+	uint32_t regval = 0;
+
+	clk_enable(tzc_dev->pdata.clk[0]);
+
+	regval = io_read32(base + BUILD_CONFIG_OFF);
+	ddata->nb_filters = ((regval >> BUILD_CONFIG_NF_SHIFT) &
+			     BUILD_CONFIG_NF_MASK) + 1;
+	ddata->nb_regions = ((regval >>	BUILD_CONFIG_NR_SHIFT) &
+			     BUILD_CONFIG_NR_MASK);
+
+	clk_disable(tzc_dev->pdata.clk[0]);
+
+	DMSG("TZC400 Filters %"PRIu32" Regions %"PRIu32, ddata->nb_filters,
+	     ddata->nb_regions);
+}
+
+static struct tzc_device *tzc_alloc(void)
+{
+	struct tzc_device *tzc_dev = NULL;
+	struct stm32mp_tzc_driver_data *ddata = NULL;
+
+	tzc_dev = calloc(1, sizeof(*tzc_dev));
+	ddata = calloc(1, sizeof(*ddata));
+
+	if (tzc_dev && ddata) {
+		tzc_dev->ddata = ddata;
+		return tzc_dev;
+	}
+
+	free(ddata);
+	free(tzc_dev);
+
+	return NULL;
+}
+
+static void tzc_free(struct tzc_device *tzc_dev)
+{
+	if (tzc_dev) {
+		free(tzc_dev->ddata);
+		free(tzc_dev);
+	}
+}
+
+static TEE_Result stm32mp_tzc_parse_fdt(struct tzc_device *tzc_dev,
+					const void *fdt, int node)
 {
-	void *base = phys_to_virt(TZC_BASE, MEM_AREA_IO_SEC, 1);
 	TEE_Result res = TEE_ERROR_GENERIC;
+	struct dt_node_info dt_info = { };
+	struct io_pa_va base = { };
+	int offs = 0;
 
-	assert(base);
-	tzc_init((vaddr_t)base);
+	fdt_fill_device_info(fdt, &dt_info, node);
+	if (dt_info.reg == DT_INFO_INVALID_REG ||
+	    dt_info.reg_size == DT_INFO_INVALID_REG_SIZE)
+		panic("Missing properties in TZC DT node");
+
+	base.pa = dt_info.reg;
+	tzc_dev->pdata.name = fdt_get_name(fdt, node, NULL);
+	tzc_dev->pdata.base = io_pa_or_va_secure(&base, dt_info.reg_size);
+	tzc_dev->pdata.irq = dt_info.interrupt;
+
+	res = clk_dt_get_by_index(fdt, node, 0, tzc_dev->pdata.clk);
+	if (res)
+		return res;
+
+	res = clk_dt_get_by_index(fdt, node, 1, tzc_dev->pdata.clk + 1);
+	if (res || !tzc_dev->pdata.clk[1])
+		DMSG("No secondary clock for %s",
+		     fdt_get_name(fdt, node, NULL));
+
+	/* Use memory node instead of that new one */
+	offs = fdt_node_offset_by_prop_value(fdt, offs, "device_type",
+					     "memory", sizeof("memory"));
+	if (offs < 0)
+		panic("No memory reference for TZC DT node");
+
+	tzc_dev->pdata.mem_base = fdt_reg_base_address(fdt, offs);
+	tzc_dev->pdata.mem_size = fdt_reg_size(fdt, offs);
+
+	assert(tzc_dev->pdata.mem_base != (paddr_t)-1 &&
+	       tzc_dev->pdata.mem_size != (size_t)-1);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32mp1_tzc_probe(const void *fdt, int node,
+				     const void *compt_data __unused)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct tzc_device *tzc_dev = NULL;
+
+	assert(fdt && node >= 0);
+
+	tzc_dev = tzc_alloc();
+	if (!tzc_dev)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	res = stm32mp_tzc_parse_fdt(tzc_dev, fdt, node);
+	if (res)
+		goto err;
+
+	tzc_set_driverdata(tzc_dev);
+	tzc_dev->reg = calloc(tzc_dev->ddata->nb_regions,
+				sizeof(*tzc_dev->reg));
+	tzc_dev->reg_locked = calloc(tzc_dev->ddata->nb_regions,
+				     sizeof(*tzc_dev->reg_locked));
+	if (!tzc_dev->reg || !tzc_dev->reg_locked) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto err;
+	}
+
+	clk_enable(tzc_dev->pdata.clk[0]);
+	if (tzc_dev->pdata.clk[1])
+		clk_enable(tzc_dev->pdata.clk[1]);
+
+	tzc_init((vaddr_t)tzc_dev->pdata.base);
 	tzc_dump_state();
 
 	stm32mp_tzc_check_boot_region();
@@ -124,5 +271,22 @@ static TEE_Result init_stm32mp1_tzc(void)
 	tzc_set_action(TZC_ACTION_INT);
 
 	return TEE_SUCCESS;
+
+err:
+	free(tzc_dev->reg);
+	tzc_free(tzc_dev);
+
+	return res;
 }
-driver_init(init_stm32mp1_tzc);
+
+static const struct dt_device_match tzc_secu_match_table[] = {
+	{ .compatible = "st,stm32mp1-tzc" },
+	{ }
+};
+
+DEFINE_DT_DRIVER(tzc_stm32mp1_dt_driver) = {
+	.name = "stm32mp1-tzc400",
+	.type = DT_DRIVER_NOTYPE,
+	.match_table = tzc_secu_match_table,
+	.probe = stm32mp1_tzc_probe,
+};
-- 
2.47.2

